File: Graph/DisOfNearCellHaving1/dis.cpp
#include <vector>
#include <queue>
using namespace std;

class Solution {
  public:
    typedef pair<pair<int,int>,int> pip;
    vector<vector<int>> nearest(vector<vector<int>>& mat) {
        // code here
        int m = mat.size();
        int n = mat[0].size();
        vector<vector<int>>dis(m,vector<int>(n,-1));
        queue<pip>qu;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(mat[i][j]){
                    qu.push({{i,j},0});
                    dis[i][j] = 0;
                }
            }
        }
        while(!qu.empty()){
            pip p = qu.front();
            qu.pop();
            int row = p.first.first;
            int col = p.first.second;
            int newDis = p.second+1;
            if(row>0 && mat[row-1][col]==0 && dis[row-1][col]==-1){
                dis[row-1][col] = newDis;
                qu.push({{row-1,col},newDis});
            }

            if(row<m-1 && mat[row+1][col]==0 && dis[row+1][col]==-1){
                dis[row+1][col] = newDis;
                qu.push({{row+1,col},newDis});
            }

            if(col<n-1 && mat[row][col+1]==0 && dis[row][col+1]==-1){
                dis[row][col+1] = newDis;
                qu.push({{row,col+1},newDis});
            }

            if(col>0 && mat[row][col-1]==0 && dis[row][col-1]==-1){
                dis[row][col-1] = newDis;
                qu.push({{row,col-1},newDis});
            }
        }
        return dis;
    }
};

File: Graph/DisOfNearCellHaving1/dis.py
from collections import deque

class Solution:
    def nearest(self, mat):
        m = len(mat)
        n = len(mat[0])

        dis = [[-1] * n for _ in range(m)]
        qu = deque()

        # Push all 1's into the queue
        for i in range(m):
            for j in range(n):
                if mat[i][j]:
                    qu.append(((i, j), 0))
                    dis[i][j] = 0

        while qu:
            (row, col), dist = qu.popleft()
            newDist = dist + 1

            if row > 0 and mat[row - 1][col] == 0 and dis[row - 1][col] == -1:
                dis[row - 1][col] = newDist
                qu.append(((row - 1, col), newDist))

            if row < m - 1 and mat[row + 1][col] == 0 and dis[row + 1][col] == -1:
                dis[row + 1][col] = newDist
                qu.append(((row + 1, col), newDist))

            if col < n - 1 and mat[row][col + 1] == 0 and dis[row][col + 1] == -1:
                dis[row][col + 1] = newDist
                qu.append(((row, col + 1), newDist))

            if col > 0 and mat[row][col - 1] == 0 and dis[row][col - 1] == -1:
                dis[row][col - 1] = newDist
                qu.append(((row, col - 1), newDist))

        return dis